include/forcing/DataProvider.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef NGEN_DATAPROVIDER_HPP
+#define NGEN_DATAPROVIDER_HPP
+
+#include <string>
+
+
+namespace data_access
+{
+    /** An abstraction for classes that provider access to data.
+    *
+    * Data may be pre-provided from some external source, internally calculated by the implementing type, or some
+    * combination of both.
+    */
+
+    template <class data_type, class selection_type> class DataProvider
+    {
+        /** This class provides a generic interface to data services
+        *
+        */
+
+        public:
+
+        virtual ~DataProvider() = default;
+
+        /** Return the variables that are accessable by this data provider */
+
+        virtual const std::veector<std::string> get_avaliable_variable_names() = 0;
+
+        /** Return the first valid time for which data from the request variaable  can be requested */
+
+        virtual get_data_start_time(std::string var) = 0;
+
+        /** Return the last valid time for which data from the requested variablle can be requested */
+
+        virtual get_data_stop_time(std::string var) = 0;
+
+        /**
+         * Get the index of the data time step that contains the given point in time.
+         *
+         * An @ref std::out_of_range exception should be thrown if the time is not in any time step.
+         *
+         * @param epoch_time The point in time, as a seconds-based epoch time.
+         * @return The index of the forcing time step that contains the given point in time.
+         * @throws std::out_of_range If the given point is not in any time step.
+         */
+        virtual size_t get_ts_index_for_time(const time_t &epoch_time) = 0;
+
+        /**
+         * Get the value of a forcing property for an arbitrary time period, converting units if needed.
+         *
+         * An @ref std::out_of_range exception should be thrown if the data for the time period is not available.
+         *
+         * @param selector Data required to establish what subset of the stored data should be accessed
+         * @param variable_name The name of the data property of interest.
+         * @param init_time_epoch The epoch time (in seconds) of the start of the time period.
+         * @param duration_seconds The length of the time period, in seconds.
+         * @param output_units The expected units of the desired output value.
+         * @return The value of the forcing property for the described time period, with units converted if needed.
+         * @throws std::out_of_range If data for the time period is not available.
+         */
+        virtual data_type get_value(selection_type selector, const std::string &variable_name, const time_t &init_time, const long &duration_s,
```

Comment:
How does this intend to deal with frequency?  Do we need to resample/interpolate/ect to align period and frequency of a data source to the requester?

Reply:
The function should resample as needed to get a record of **data_type**. For cases where a sequences of values is intended to be retrieved **data_type** can be a type the incapsulates and array with desired length or it can be multi-dimensional matrix for retrieval of data cubes.

Reply:
In the example netcdf implementation **data_type** is double so the intent would be to retrieve one value. If the time range was sufficiently large for the file to contain multiple values they should be merged and if it does not span a full file time step a partial result should be returned.

Reply:
Why separate `variable_name`, `init_time`, and `duration_s` from `selector`?

Reply:
That was a an attempt to make this API a close to the existing API as possible.